internal/isogeny/curve_ops.go

```diff
@@ -0,0 +1,440 @@
+package internal
+
+type CurveOperations struct {
+    Params *SidhParams
+}
+
+// Computes j-invariant for a curve y2=x3+A/Cx+x with A,C in F_(p^2). Result
+// is returned in jBytes buffer, encoded in little-endian format. Caller
+// provided jBytes buffer has to be big enough to j-invariant value. In case
+// of SIDH, buffer size must be at least size of shared secret.
+// Implementation corresponds to Algorithm 9 from SIKE.
+func (c *CurveOperations) Jinvariant(cparams *ProjectiveCurveParameters, jBytes []byte) {
+    var j, t0, t1 Fp2Element
+
+    op := c.Params.Op
+    op.Square(&j, &cparams.A)  // j  = A^2
+    op.Square(&t1, &cparams.C) // t1 = C^2
+    op.Add(&t0, &t1, &t1)      // t0 = t1 + t1
+    op.Sub(&t0, &j, &t0)       // t0 = j - t0
+    op.Sub(&t0, &t0, &t1)      // t0 = t0 - t1
+    op.Sub(&j, &t0, &t1)       // t0 = t0 - t1
+    op.Square(&t1, &t1)        // t1 = t1^2
+    op.Mul(&j, &j, &t1)        // j = j * t1
+    op.Add(&t0, &t0, &t0)      // t0 = t0 + t0
+    op.Add(&t0, &t0, &t0)      // t0 = t0 + t0
+    op.Square(&t1, &t0)        // t1 = t0^2
+    op.Mul(&t0, &t0, &t1)      // t0 = t0 * t1
+    op.Add(&t0, &t0, &t0)      // t0 = t0 + t0
+    op.Add(&t0, &t0, &t0)      // t0 = t0 + t0
+    op.Inv(&j, &j)             // j  = 1/j
+    op.Mul(&j, &t0, &j)        // j  = t0 * j
+
+    c.Fp2ToBytes(jBytes, &j)
+}
+
+// Given affine points x(P), x(Q) and x(Q-P) in a extension field F_{p^2}, function
+// recorvers projective coordinate A of a curve. This is Algorithm 10 from SIKE.
+func (c *CurveOperations) RecoverCoordinateA(curve *ProjectiveCurveParameters, xp, xq, xr *Fp2Element) {
+    var t0, t1 Fp2Element
+
+    op := c.Params.Op
+    op.Add(&t1, xp, xq)                            // t1 = Xp + Xq
+    op.Mul(&t0, xp, xq)                            // t0 = Xp * Xq
+    op.Mul(&curve.A, xr, &t1)                      // A  = X(q-p) * t1
+    op.Add(&curve.A, &curve.A, &t0)                // A  = A + t0
+    op.Mul(&t0, &t0, xr)                           // t0 = t0 * X(q-p)
+    op.Sub(&curve.A, &curve.A, &c.Params.OneFp2)   // A  = A - 1
+    op.Add(&t0, &t0, &t0)                          // t0 = t0 + t0
+    op.Add(&t1, &t1, xr)                           // t1 = t1 + X(q-p)
+    op.Add(&t0, &t0, &t0)                          // t0 = t0 + t0
+    op.Square(&curve.A, &curve.A)                  // A  = A^2
+    op.Inv(&t0, &t0)                               // t0 = 1/t0
+    op.Mul(&curve.A, &curve.A, &t0)                // A  = A * t0
+    op.Sub(&curve.A, &curve.A, &t1)                // A  = A - t1
+}
+
+// Computes equivalence (A:C) ~ (A+2C : A-2C)
+func (c *CurveOperations) CalcCurveParamsEquiv3(cparams *ProjectiveCurveParameters) CurveCoefficientsEquiv {
```

Comment:
Because it computes curve equivalent to the one described by cparam. Conversion is done in order to improve performance of algorithms doing operations on those curves. As it's projective P^1 space, it doesn't matter which curve exactly is used as long as it's equivalent to the one described by cparam. Meaning of coef.A and coef.C is described in a comment for ``CurveCoefficientsEquiv3``. Nevertheless name of a struct is a bit long.